README.rst
.. raw:: html

    <img src="docs/images/logo-120.png" width="120px">
    <br/>

Petastorm
=========

.. image:: https://travis-ci.com/uber/petastorm.svg?branch=master
   :target: https://travis-ci.com/uber/petastorm
   :alt: Build Status (Travis CI)

.. image:: https://codecov.io/gh/uber/petastorm/branch/master/graph/badge.svg
   :target: https://codecov.io/gh/uber/petastorm/branch/master
   :alt: Code coverage

.. image:: https://img.shields.io/badge/License-Apache%202.0-blue.svg
   :target: https://img.shields.io/badge/License-Apache%202.0-blue.svg
   :alt: License

.. image:: https://badge.fury.io/py/petastorm.svg
   :target: https://pypi.org/project/petastorm
   :alt: Latest Version

.. inclusion-marker-start-do-not-remove

Petastorm is an open source data access library developed at Uber ATG. This library enables single machine or
distributed training and evaluation of deep learning models directly from datasets in Apache Parquet
format. Petastorm supports popular Python-based machine learning (ML) frameworks such as
`Tensorflow <http://www.tensorflow.org/>`_, `PyTorch <https://pytorch.org/>`_, and
`PySpark <http://spark.apache.org/docs/latest/api/python/pyspark.html>`_. It can also be used from pure Python code.

Documentation web site: `<https://petastorm.readthedocs.io>`_


Installation
------------

.. code-block:: bash

    pip install petastorm


There are several extra dependencies that are defined by the ``petastorm`` package that are not installed automatically.
The extras are: ``tf``, ``tf_gpu``, ``torch``, ``opencv``, ``docs``, ``test``.

For example to trigger installation of GPU version of tensorflow and opencv, use the following pip command:

.. code-block:: bash

    pip install petastorm[opencv,tf_gpu]



Generating a dataset
--------------------

A dataset created using Petastorm is stored in `Apache Parquet <https://parquet.apache.org/>`_ format.
On top of a Parquet schema, petastorm also stores higher-level schema information that makes multidimensional arrays into a native part of a petastorm dataset. 

Petastorm supports extensible data codecs. These enable a user to use one of the standard data compressions (jpeg, png) or implement her own.

Generating a dataset is done using PySpark.
PySpark natively supports Parquet format, making it easy to run on a single machine or on a Spark compute cluster.
Here is a minimalistic example writing out a table with some random data.


.. code-block:: python

    HelloWorldSchema = Unischema('HelloWorldSchema', [
       UnischemaField('id', np.int32, (), ScalarCodec(IntegerType()), False),
       UnischemaField('image1', np.uint8, (128, 256, 3), CompressedImageCodec('png'), False),
       UnischemaField('other_data', np.uint8, (None, 128, 30, None), NdarrayCodec(), False),
    ])


    def row_generator(x):
       """Returns a single entry in the generated dataset. Return a bunch of random values as an example."""
       return {'id': x,
               'image1': np.random.randint(0, 255, dtype=np.uint8, size=(128, 256, 3)),
               'other_data': np.random.randint(0, 255, dtype=np.uint8, size=(4, 128, 30, 3))}


    def generate_hello_world_dataset(output_url='file:///tmp/hello_world_dataset'):
       rows_count = 10
       rowgroup_size_mb = 256

       spark = SparkSession.builder.config('spark.driver.memory', '2g').master('local[2]').getOrCreate()
       sc = spark.sparkContext

       # Wrap dataset materialization portion. Will take care of setting up spark environment variables as
       # well as save petastorm specific metadata
       with materialize_dataset(spark, output_url, HelloWorldSchema, rowgroup_size_mb):

           rows_rdd = sc.parallelize(range(rows_count))\
               .map(row_generator)\
               .map(lambda x: dict_to_spark_row(HelloWorldSchema, x))

           spark.createDataFrame(rows_rdd, HelloWorldSchema.as_spark_schema()) \
               .coalesce(10) \
               .write \
               .mode('overwrite') \
               .parquet(output_url)

- ``HelloWorldSchema`` is an instance of a ``Unischema`` object.
  ``Unischema`` is capable of rendering types of its fields into different
  framework specific formats, such as: Spark ``StructType``, Tensorflow
  ``tf.DType`` and numpy ``numpy.dtype``.
- To define a dataset field, you need to specify a ``type``, ``shape``, a
  ``codec`` instance and whether the field is nullable for each field of the
  ``Unischema``.
- We use PySpark for writing output Parquet files. In this example, we launch
  PySpark on a local box (``.master('local[2]')``). Of course for a larger
  scale dataset generation we would need a real compute cluster.
- We wrap spark dataset generation code with the ``materialize_dataset``
  context manager.  The context manager is responsible for configuring row
  group size at the beginning and write out petastorm specific metadata at the
  end.
- The row generating code is expected to return a Python dictionary indexed by
  a field name. We use ``row_generator`` function for that. 
- ``dict_to_spark_row`` converts the dictionary into a ``pyspark.Row``
  object while ensuring schema ``HelloWorldSchema`` compliance (shape,
  type and is-nullable condition are tested).
- Once we have a ``pyspark.DataFrame`` we write it out to a parquet
  storage. The parquet schema is automatically derived from
  ``HelloWorldSchema``.

Plain Python API
----------------
The ``petastorm.reader.Reader`` class is the main entry point for user
code that accesses the data from an ML framework such as Tensorflow or Pytorch.
The reader has multiple features such as:

- Selective column readout
- Multiple parallelism strategies: thread, process, single-threaded (for debug)
- N-grams readout support
- Row filtering (row predicates)
- Shuffling
- Partitioning for multi-GPU training
- Local caching

Reading a dataset is simple using the ``petastorm.reader.Reader`` class which can be created using the
``petastorm.make_reader`` factory method:

.. code-block:: python

   from petastorm import make_reader

    with make_reader('hdfs://myhadoop/some_dataset') as reader:
       for row in reader:
           print(row)

``hdfs://...`` and ``file://...`` are supported URL protocols.

Once a ``Reader`` is instantiated, you can use it as an iterator.

Tensorflow API
--------------

To hookup the reader into a tensorflow graph, you can use the ``tf_tensors``
function:

.. code-block:: python

    with make_reader('file:///some/localpath/a_dataset') as reader:
       row_tensors = tf_tensors(reader)
       with tf.Session() as session:
           for _ in range(3):
               print(session.run(row_tensors))

Alternatively, you can use new ``tf.data.Dataset`` API;

.. code-block:: python

    with make_reader('file:///some/localpath/a_dataset') as reader:
        dataset = make_petastorm_dataset(reader)
        iterator = dataset.make_one_shot_iterator()
        tensor = iterator.get_next()
        with tf.Session() as sess:
            sample = sess.run(tensor)
            print(sample.id)

Pytorch API
-----------

As illustrated in
`pytorch_example.py <https://github.com/uber/petastorm/blob/master/examples/mnist/pytorch_example.py>`_,
reading a petastorm dataset from pytorch
can be done via the adapter class ``petastorm.pytorch.DataLoader``,
which allows custom pytorch collating function and transforms to be supplied.

Be sure you have ``torch`` and ``torchvision`` installed:

.. code-block:: bash

    pip install torchvision

The minimalist example below assumes the definition of a ``Net`` class and
``train`` and ``test`` functions, included in ``pytorch_example``:

.. code-block:: python

    import torch
    from petastorm.pytorch import DataLoader

    torch.manual_seed(1)
    device = torch.device('cpu')
    model = Net().to(device)
    optimizer = torch.optim.SGD(model.parameters(), lr=0.01, momentum=0.5)

    def _transform_row(mnist_row):
        transform = transforms.Compose([
            transforms.ToTensor(),
            transforms.Normalize((0.1307,), (0.3081,))
        ])
        return (transform(mnist_row['image']), mnist_row['digit'])

    with DataLoader(make_reader('file:///localpath/mnist/train', num_epochs=10),
                    batch_size=64, transform=_transform_row) as train_loader:
        train(model, device, train_loader, 10, optimizer, 1)
    with DataLoader(make_reader('file:///localpath/mnist/test', num_epochs=10),
                    batch_size=1000, transform=_transform_row) as test_loader:
        test(model, device, test_loader)

PySpark and SQL
---------------

Using the Parquet data format, which is natively supported by Spark, makes it possible to use a wide range of Spark
tools to analyze and manipulate the dataset. The example below shows how to read a Petastorm dataset
as a Spark RDD object:

.. code-block:: python

   # Create a dataframe object from a parquet file
   dataframe = spark.read.parquet(dataset_url)

   # Show a schema
   dataframe.printSchema()

   # Count all
   dataframe.count()

   # Show a single column
   dataframe.select('id').show()

SQL can be used to query a Petastorm dataset:

.. code-block:: python

   spark.sql(
      'SELECT count(id) '
      'from parquet.`file:///tmp/hello_world_dataset`').collect()

You can find a full code sample here: `pyspark_hello_world.py <https://github.com/uber/petastorm/blob/master/examples/hello_world/pyspark_hello_world.py>`_,

.. inclusion-marker-end-do-not-remove
   Place contents above here if they should also appear in read-the-docs.
   Contents below are already part of the read-the-docs table of contents.

Troubleshooting
---------------

See the Troubleshooting_ page and please submit a ticket_ if you can't find an
answer.


Development
-----------

See the Development_ page for instructions on how to develop Petastorm and
run tests.


Publications
------------

1. Gruener, R., Cheng, O., and Litvin, Y. (2018) *Introducing Petastorm: Uber ATG's Data Access Library for Deep Learning*. URL: https://eng.uber.com/petastorm/

.. raw:: html

    <div style="position: relative; padding-bottom: 56.25%; height: 0; overflow: hidden; max-width: 100%; height: auto;">
        <iframe src="//www.youtube.com/embed/CcGVCjqXgUY" frameborder="0" allowfullscreen style="position: absolute; top: 0; left: 0; width: 100%; height: 100%;"></iframe>
    </div>

.. _Troubleshooting: docs/troubleshoot.rst
.. _ticket: https://github.com/uber/petastorm/issues/new
.. _Development: docs/development.rst


docs/release-notes.rst
.. inclusion-marker-start-do-not-remove

=============
Release notes
=============

Release 0.9.9 (unreleased)
==========================
- `PR 640 <https://github.com/uber/petastorm/pull/640>`_: Security fix for arbitrary code execution (affects depickling of Unischema loaded from petastorm datasets).


Release 0.9.8
==========================
Thanks to our new contributor - Lionel Gueguen, Ritwik Bera, acmore and Fardin Abdi!

- `PR 617 <https://github.com/uber/petastorm/pull/617>`_: Deprecating `pyarrow_serialize` argument of `petastorm.make_reader`.
- `PR 625 <https://github.com/uber/petastorm/pull/625>`_: Support on-premise s3-compatible storage. User can now pass an instance of pyarrow filesystem to `make_reader` and `make_batch_reader` functions.
- `PR 555 <https://github.com/uber/petastorm/pull/555>`_: Support in-memory caching of data in the PyTorch BatchedDataLoader.


Release 0.9.7
==========================

- `PR 611 <https://github.com/uber/petastorm/pull/611>`_: Bugfix: S3FSWrapper is deprecated at s3fs 0.5.0. Minimal s3fs version required by petastorm wheel is now 0.5.0.
- `PR 615 <https://github.com/uber/petastorm/pull/615>`_: Raising an error if an empty list of columns is selected by a user.
- `PR 614 <https://github.com/uber/petastorm/pull/615>`_: Minimal supported version of pyarrow is now 0.17.1.
- `PR 616 <https://github.com/uber/petastorm/pull/616>`_: Added support for np.uint8.


Release 0.9.6
==========================
Thanks to our new contributor - David McGuire!

- `PR 588 <https://github.com/uber/petastorm/pull/588>`_: ``make_reader`` can now open a parquet dataset from a subdirectory in an s3 bucket.
- `PR 594 <https://github.com/uber/petastorm/pull/594>`_: Parameterize factory methods with s3 configs
- `PR 596 <https://github.com/uber/petastorm/pull/596>`_: Add a flag to factory methods to allow zmq copy buffers to be disabled
- `PR 601 <https://github.com/uber/petastorm/pull/601>`_: Allow user to use s3, s3a and s3n url schemes when writing datasets.


Release 0.9.5
==========================
- Removed gcsfs mandatory dependency from the python wheel.


Release 0.9.4
==========================

- Fixed Python3.6 compatibility issue of UnischemaField namedtuple class.


Release 0.9.3
==========================

Thanks to our new contributors: Travis Addair and Ryan (rb-determined-ai).

- Retire support for Python 2.
- `PR 568 <https://github.com/uber/petastorm/pull/568>`_: Added additional kwargs for Spark Dataset Converter.
- `PR 564 <https://github.com/uber/petastorm/pull/564>`_: Expose filters (PyArrow filters) argument in make_reader and make_batch_reader



Release 0.9.2
=============
- `PR 556 <https://github.com/uber/petastorm/pull/556>`_: Remove driver param for hdfs.connect when using pyarrow 0.17 and above.


Release 0.9.1
=============

Bug fixes and additional features
---------------------------------
- `PR 554 <https://github.com/uber/petastorm/pull/554>`_: Fixed ``make_petastorm_dataset`` compatibility issue with ``WeightedSamplingReader``.
- `PR 535 <https://github.com/uber/petastorm/pull/535>`_: Stip trailing ``/`` from directory names when using ``GCSFSWrapper``'s ``walk``.


Release 0.9.0
=============

Thanks to our contributors: Weichen Xu, Liang Zhang, Felipe Petroski Such, Jennifer Blight and Greg Arutiunian!

Major features
--------------

- Apache PySpark 3.0 (Preview) support
- TensorFlow 2 support
- PR 540: petastorm.pytorch.BatchedDataLoader is a much faster version of petastorm.pytorch.DataLoader. petastorm.pytorch.BatchedDataLoader support only types that can be converted to pytorch tensors.
- Spark Dataset Converter: Spark converter API simplifies the data conversion from Spark to TensorFlow or PyTorch. The input Spark DataFrame is first materialized in the parquet format and then loaded as a tf.data.Dataset or torch.utils.data.DataLoader.

Bug fixes and additional features
---------------------------------
- `PR 513 <https://github.com/uber/petastorm/pull/513>`_: Preserve spark dataframe schema order in a namedtuples produced by `make_reader` and `make_batch_reader`. A backward compatiility flag is `here <https://github.com/uber/petastorm/blob/63c1faf9de3ebc56a386117bdd68deccd896fe25/petastorm/unischema.py#L37>`_.
- `PR 490 <https://github.com/uber/petastorm/pull/490>`_: TransformSpecs was not being applied if there was also a row predicate specified when using make_reader.
- `PR 495 <https://github.com/uber/petastorm/pull/495>`_: Removed spurious warning: “schema_fields behavior has changed. Now, regular expression pattern must match the entire field name...the following fields: “
- `PR 498 <https://github.com/uber/petastorm/pull/498>`_: Raise an explicit error if a non-scalar value is used as a value for a scalar field.
- `PR 529 <https://github.com/uber/petastorm/pull/529>`_: Using parquet.summary.metadata.level to control summary metadata file creation. This eliminates depreciation warning.
- `PR 533 <https://github.com/uber/petastorm/pull/533>`_:  make_petastorm_dataset (TF tf.data API) function supports ngrams.
- `PR 544 <https://github.com/uber/petastorm/pull/544>`_: Enable multiple iterations over the same petastorm.pytorch.DataLoader instance.

Python 2 support is deprecated and it will be removed in the next release.


Release 0.8.2
=============

New features and bug fixes
--------------------------
- pandas 1.0 compatibility issue resolved.

Release 0.8.1
=============

Thanks to our new contributor: Sergey Serebryakov for contributing Google Cloud Storage (GCS) filesystem implementation.

New features and bug fixes
--------------------------
- Bug fix: when ``make_batch_reader`` was used and Parquet files in a dataset have different number of row-groups. Symptoms:

  - ``pyarrow.lib.ArrowIOError: The file only has <X> row groups, requested metadata for row group: <Y>`` (where <X> and <Y> are integers) error message;
  - loss of some rows in a dataset.


examples/mnist/main.py
#  Copyright (c) 2017-2018 Uber Technologies, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

###
# Adapted to petastorm dataset using original contents from
# https://github.com/pytorch/examples/mnist/main.py .
###
from __future__ import print_function
import argparse
import os
import sys
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torch.utils.data.dataloader import default_collate
from torchvision import transforms

from petastorm.reader import Reader, BatchReader
from petastorm.workers_pool.thread_pool import ThreadPool


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
        self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
        self.conv2_drop = nn.Dropout2d()
        self.fc1 = nn.Linear(320, 50)
        self.fc2 = nn.Linear(50, 10)

    def forward(self, x):
        x = F.relu(F.max_pool2d(self.conv1(x), 2))
        x = F.relu(F.max_pool2d(self.conv2_drop(self.conv2(x)), 2))
        x = x.view(-1, 320)
        x = F.relu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        x = self.fc2(x)
        return F.log_softmax(x, dim=1)

def train(args, model, device, train_loader, optimizer, epoch):
    model.train()
    for batch_idx, batch in enumerate(train_loader):
        (data, target) = default_collate(batch)
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        if batch_idx % args.log_interval == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                epoch, batch_idx * len(data), len(train_loader.reader),
                100. * batch_idx / len(train_loader), loss.item()))

def test(args, model, device, test_loader, epoch):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for batch in test_loader:
            (data, target) = default_collate(batch)
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item() # sum up batch loss
            pred = output.max(1, keepdim=True)[1] # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.reader)
    print('\nTest Set Epoch {}: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        epoch, test_loss, correct, len(test_loader.reader),
        100. * correct / len(test_loader.reader)))

def main():
    # Training settings
    parser = argparse.ArgumentParser(description='Petastorm MNIST Example')
    parser.add_argument('--dataset-url', type=str,
                        default='file:///{}'.format(os.path.abspath(os.path.join(
                            os.path.dirname(sys.argv[0]), '..', '..', '..'))),
                        metavar='S',
                        help='hdfs:// or file:/// URL to the MNIST petastorm dataset')
    parser.add_argument('--batch-size', type=int, default=64, metavar='N',
                        help='input batch size for training (default: 64)')
    parser.add_argument('--test-batch-size', type=int, default=1000, metavar='N',
                        help='input batch size for testing (default: 1000)')
    parser.add_argument('--epochs', type=int, default=10, metavar='N',
                        help='number of epochs to train (default: 10)')
    parser.add_argument('--lr', type=float, default=0.01, metavar='LR',
                        help='learning rate (default: 0.01)')
    parser.add_argument('--momentum', type=float, default=0.5, metavar='M',
                        help='SGD momentum (default: 0.5)')
    parser.add_argument('--no-cuda', action='store_true', default=False,
                        help='disables CUDA training')
    parser.add_argument('--seed', type=int, default=1, metavar='S',
                        help='random seed (default: 1)')
    parser.add_argument('--log-interval', type=int, default=10, metavar='N',
                        help='how many batches to wait before logging training status')
    args = parser.parse_args()
    use_cuda = not args.no_cuda and torch.cuda.is_available()

    torch.manual_seed(args.seed)

    device = torch.device("cuda" if use_cuda else "cpu")

    model = Net().to(device)
    optimizer = optim.SGD(model.parameters(), lr=args.lr, momentum=args.momentum)

    def transform_row(mnist_row):
        transform = transforms.Compose([
            transforms.Lambda(lambda nd: nd.reshape(28, 28, 1)),
            transforms.ToTensor(),
            transforms.Normalize((0.1307,), (0.3081,))
        ])
        return (transform(mnist_row.image), mnist_row.digit)

    for epoch in range(1, args.epochs + 1):
        # Handle multiple epochs by reinstantiating reader per epoch
        with Reader('{}/train'.format(args.dataset_url), shuffle=True, reader_pool=ThreadPool(1), num_epochs=1) as reader:
            train(args, model, device, BatchReader(reader, args.batch_size, transform=transform_row), optimizer, epoch)

        with Reader('{}/test'.format(args.dataset_url), shuffle=True, reader_pool=ThreadPool(1), num_epochs=1) as reader:
            test(args, model, device, BatchReader(reader, args.batch_size, transform=transform_row), epoch)


if __name__ == '__main__':
    main()
